### ISSUE TITLE
Spec: minimum implementation requirements for stub problems (v1)

### ISSUE THREAD

**Issue by 0bserver07** [cybertronai/hinton-problems#1]
# Minimum implementation spec for Hinton problem stubs (v1)

## Purpose

This catalog has 53 problem stubs and one worked example ([`encoder-4-2-4/`](../tree/main/encoder-4-2-4)). To dispatch parallel agents on the remaining stubs, we need a uniform shape so:

- Implementations are comparable across problems
- Results land in a predictable place
- Reviewers can validate without reading 52 different layouts

This issue is the source of truth for "minimum" — what every stub must contain to be considered done. Per-stub implementation issues link back here.

The shape is derived from two existing worked examples:

- [`encoder-4-2-4/`](../tree/main/encoder-4-2-4) — this repo
- [`wip-boltzmann-shifter/`](https://github.com/cybertronai/sutro-problems/tree/main/wip-boltzmann-shifter) — sister repo, same shape

## Required files

| File | Purpose |
|---|---|
| `<name>.py` (or `problem.py`) | Dataset generation + model + train + eval |
| `README.md` | 8-section template below |
| `make_<name>_gif.py` | Generates the animation at the top of the README |
| `visualize_<name>.py` | Training curves + final weights / codes |
| `<name>.gif` | Committed animation |
| `viz/` | Committed PNGs (training curves, weight matrix, etc.) |

## README sections

1. **Header** — source paper citation (full), one-sentence "demonstrates", inline GIF
2. **Problem** — inputs/outputs, training distribution, *the interesting property* (one paragraph: why this problem is non-trivial)
3. **Files** — table mapping each file to its purpose
4. **Running** — one-line command, expected wall-clock, expected accuracy
5. **Results** — table with: final accuracy, per-class breakdown if multi-class, training time, full hyperparameters used
6. **Visualizations** — each PNG inlined, with prose explaining what the reader is looking at
7. **Deviations from the original procedure** — honest list (e.g., CD-k vs simulated annealing, modern hardware, different hidden-unit count). If faithful, write "None — faithful reproduction."
8. **Open questions / next experiments** — at least one

## Reproducibility

- Seed exposed via CLI flag (`--seed`)
- All hyperparameters appear in the Results table
- Final reported number reproduces by re-running the command in §Running

## Acceptance checklist

A stub is "minimum-spec done" when:

- [ ] Reproduces in under 5 minutes on a laptop (no GPU required at v1)
- [ ] Final accuracy / metric reported with seed in the Results table
- [ ] GIF illustrates **both** the problem definition AND the learning dynamics
- [ ] At least one weight visualization (Hinton-diagram or heatmap)
- [ ] At least one training-curves plot
- [ ] "Deviations from the original" section present
- [ ] At least one entry in "Open questions / next experiments"
- [ ] No `NotImplementedError` left in the implementation file

## Out of scope for v1

- **Energy metric** (ByteDMD / reuse distance) is the eventual goal of the broader Sutro effort, but v1 only requires correctness + visualization. Energy-efficient variants will be a follow-up pass once baselines exist.
- **GPU / large-scale runs** — every problem here is a 1980s/1990s toy. A laptop should suffice.

## How per-stub issues will use this spec

Each per-stub issue will:

1. Link back to this issue
2. Cite the source paper (already in the stub's `README.md`)
3. Be picked up by an agent that follows the checklist above

Reviewers use the checklist directly when reviewing PRs.

## Revisions

This is v1. If a stub reveals the spec is wrong or insufficient, comment here and we'll iterate. Future versions will tighten correctness requirements and add the energy metric.

---

_agent-0bserver07 (Claude Code) on behalf of Yad_

**Comment by yaroslavvb** [comment]
Here are some underlying motivations. We're interested in figuring out directions for learning and an alternative approach to learning. The field has standardized on backprop by the end of the '80s, and Hinton gives a sample of problems that were used at the time.


In the last 20 years, we have transitioned to GPUs, and the math has changed considerably. Instead of being bottlenecked by arithmetic, the shrinking of transistors means that arithmetic is essentially free, and all of the work comes from data movement. Backprop is inefficient in terms of "commute to compute ratio" because it requires fetching all of the activations for each gradient add.


So a natural experiment would be to redo key experiments of this time with a focus on data movement. To this effect, the first step (current task) would be to get a baseline ... so we want to establish the list of problems which are famous (made by Hinton), reasonable to implement and are easy to run/reproduce. So the goal of this step is to create baselines of implementations and find the list of problems which are implementable by implementing them.


The next stage, out of scope at the time currently, but specifying here for reference, the next step would be to instrument the execution environment to track data movement and search for more efficient solutions to this problem in terms of the data movement. We might instrument the code to track reuse distance (ie, bytedmd, https://github.com/cybertronai/ByteDMD ) and ask to provide more efficient solutions, or for full robustness, ask for solutions in the simplified dally model -- https://github.com/cybertronai/simplified-dally-model




For each problem in the list and implementation, I want to know some metric of complexity, such as how hard it was to implement, possibly token count. I want to know the end-to-end runtime needed to rerun the current baseline implementation. This information will be used as a filtering step to select better problems for the next stage.


Ideally, I would see the following table with following columns for each problem:
1. Problem name/link
2. Able to reproduce paper (yes/no)
3. Difficulty (ie, token count or agent runtime)
4. Wallclock time of running the final experiment

**Comment by 0bserver07** [comment]
## Update from @yaroslavvb's framing — adding the v1 baseline metrics

Folding @yaroslavvb's motivation into the spec so it's visible to every agent that picks up a per-stub issue. **Original spec (above) still stands** — this comment adds three things on top.

### Why this work matters (one paragraph for context)

The field standardized on backprop by the end of the '80s. In the 20 years since, GPUs shifted the bottleneck: arithmetic is essentially free, **data movement** is where the cost lives. Backprop has a bad "commute-to-compute ratio" — every gradient step refetches all activations.

The long arc: redo Hinton's classic experiments with a focus on data movement, and search for solvers that beat backprop on that axis (ByteDMD / [simplified-dally-model](https://github.com/cybertronai/simplified-dally-model)).

**v1 (this work) is the baseline step.** We need to know which problems from the catalog are *actually* implementable and easy to reproduce, so v2 has a filtered list to attack with the data-movement instrumentation.

### New per-problem metrics (required for v1 acceptance)

Every per-stub PR must report 4 numbers, used as filter criteria for v2:

| # | Metric | How to capture |
|---:|---|---|
| 1 | Problem name + folder link | (already in catalog) |
| 2 | **Reproduces paper?** (yes / no) | Compare your final accuracy/metric to the paper's reported number. State both. "yes" if within reasonable tolerance; "no" if the paper's number isn't reachable with the documented procedure. |
| 3 | **Difficulty** | Agent token count (preferred) or agent wallclock to implement |
| 4 | **Wallclock** | Time to *run* the final experiment (`time python3 <slug>.py`) — not implementation time |

These go in the **PR description** (so they're discoverable on the merge commit) and as a **new row in the catalog table** in `README.md` (single source of truth — once all PRs land, the README is the v2 filter).

### Updated acceptance checklist (additions)

On top of the original 8 boxes:

- [ ] PR description states **"Paper reports X; we got Y. Reproduces: yes/no."**
- [ ] PR description states **wallclock to run the final experiment**
- [ ] PR description states **agent token count or agent wallclock** (whatever the harness logs)
- [ ] `README.md` catalog row updated with the 4 metrics columns

### Catalog table extension

The existing `README.md` has a year-grouped catalog. We'll extend each row to:

```
| Problem | Source paper (year) | Reproduces? | Difficulty | Run wallclock |
```

First per-stub PR to land (likely `xor`) sets the column format; subsequent PRs append rows. The main agent (or whoever lands the wave) keeps the table sorted.

### Wave plan

Implementation will roll out in supervised waves, smallest-first:

- **Wave 0 (validation, 1 stub)**: `xor` — prove the loop end-to-end, debug the metrics-reporting flow
- **Wave 1+ (parallel)**: tier-A trivial MLPs first, then 1980s/90s classics, then capsule-era / harder problems
- Each wave reviewed before the next; spec revised here as a comment if a wave reveals a gap

---

_agent-0bserver07 (Claude Code) on behalf of Yad_

**Comment by 0bserver07** [comment]
## Spec v2 — clarifications from wave 0+1 review

After reviewing the 4 wave 0+1 PRs (#3 xor, #7 symmetry, #8 n-bit-parity, #9 negation) and auditing the remaining stubs, three clarifications and one scope note.

### 1. Stub-architecture deviations

The `negation` stub specified `4-3-3`, but each output is a 2D XOR requiring ≥2 hidden units per output (6 minimum). PR #9 correctly deviated to `4-6-3` with a sweep showing `4-3-3` fails 30/30 seeds.

**New rule** (codifying what #9 already did right):

If a stub's architecture provably cannot converge:

- Run a sweep ≥30 seeds at the original arch, document the failure
- Propose a justified alternative (cite the mathematical reason, e.g., "each output is a 2D XOR needing ≥2 hidden units")
- Document both in the README's `Deviations from the original` section
- Mention the deviation in the PR body

### 2. Perturb-on-plateau is now an explicit optional checkbox

Three of four wave 0+1 PRs deferred the perturb-on-plateau wrapper. Wave 2 (`encoder-4-3-4`, `encoder-8-3-8`) implemented variants of it and recovered substantially better solve rates — `encoder-8-3-8` matched the paper's 16/20 = 80% exactly.

**Adding to acceptance checklist** (optional):

- [ ] Perturb-on-plateau wrapper (recommended if the original paper used one — check the source)

If skipped, document as "intentional v1 simplification" under Deviations. The spec doesn't *require* it, but the comparability hit is real.

### 3. External-dataset stubs deferred to v1.5

The audit identified stubs requiring external datasets (MNIST, CIFAR-10, smallNORB, sunspot data, Peterson-Barney). For v1, these are out of scope:

`smallnorb-novel-viewpoint`, `sunspots`, `vowel-mixture-experts`, `ff-hybrid-mnist`, `ff-label-in-input`, `ff-recurrent-mnist`, `ff-cifar-locally-connected`, `multimnist-capsnet`, `affnist`, `distillation-mnist-omitted-3`, `mnist-2x5-subclass`

Will revisit once v1 has solid synthetic-data baselines and a download/cache strategy.

### 4. Stubs needing spec clarification before agent dispatch

The audit flagged 8 stubs whose READMEs are too ambiguous to implement directly without an arbitrary judgment call from the agent:

`deep-lambertian-spheres`, `transforming-autoencoders`, `catch-game`, `constellations`, `geo-flow-capsules`, `multi-level-glimpse-mnist`, `distillation-mnist-omitted-3`, `affnist`

Each will get a follow-up issue with the specific clarification needed (e.g., "specify `K` as fixed or variable" for `constellations`, "define reward function" for `catch-game`).

### What's safe to fire immediately

The audit cleared **24 stubs** as implementable as-is. Wave 3 picks 6 of them: `encoder-40-10-40`, `shifter`, `grapheme-sememe`, `distributed-to-local-bottleneck`, `t-c-discrimination`, `recurrent-shift-register`.

---

_agent-0bserver07 (Claude Code) on behalf of Yad_

**Comment by yaroslavvb** [comment]
For examples using MNIST dependency, it is fine to use `from torchvision import datasets`,  `datasets.MNIST` pattern, we can assume Torchvision is installed.

**Comment by 0bserver07** [comment]
## v1 shipped — 2026-05-03

All 53 stub implementations landed in `main` per this spec. Closing the issue.

### Wave PRs (all merged)

- #32 wave 0+1 — tier-A backprop foundations (4 stubs)
- #33 wave 2 — encoder family + binary-addition (5)
- #34 wave 3 — tier-B 1980s foundational (6)
- #35 wave 4 — tier-B 1980s-90s classics (6)
- #36 wave 5 — Helmholtz / MDL / Imax / fast-weights (6)
- #37 wave 6 — TRBM / RTRBM / gated-RBM / RNN / factorial-VQ / eGLOM (6)
- #38 wave 7 — MNIST cluster: FF + distillation + capsule precursor (6)
- #39 wave 8 — external-data + harder architectures (6)
- #40 wave 9 — final hard stubs: Lambertian, catch-game, geo-flow, affnist, MultiMNIST CapsNet (5)
- #41 wave 10 — AIR + matrix capsules: air-multimnist, air-3d-primitives, smallnorb-novel-viewpoint (3)
- #42 docs — RESULTS.md catalog + v1-complete banner

### Outcome

- **53 of 53 stubs implemented** (100%)
- 27 reproduce paper claims (full or qualitative); 25 partial with documented gaps; 1 non-replication with three-cause analysis
- Pure numpy + matplotlib throughout, all stubs run on a laptop M-series CPU
- Per-stub metrics catalog: [`RESULTS.md`](https://github.com/cybertronai/hinton-problems/blob/main/RESULTS.md)

### MNIST loader note

Wave 7-10 stubs that needed MNIST predate the torchvision-allowance comment — they all use a numpy-only loader (`urllib` + `gzip` + manual parse, cached under `~/.cache/hinton-mnist/`). Works without a torch dependency. For v1.5/v2 scale-ups (paper-config FF runs, MultiMNIST CapsNet at 60M-pair scale, etc.), the `from torchvision import datasets; datasets.MNIST` pattern is the lighter ergonomic choice.

### What's next (out of scope for this issue)

- **v2 — ByteDMD / data-movement instrumentation**: the actual research goal. Per-stub reuse-distance tracking on the v1 baselines; the v2 filter section in `RESULTS.md` recommends the cleanest-converging stubs as starting points.
- **v1.5 — paper-scale re-runs** for the partial reproductions where compute (not implementation) was the gap: `ff-hybrid-mnist`, `ff-label-in-input`, `ff-recurrent-mnist`, `ff-cifar-locally-connected`, `multimnist-capsnet`, `affnist`. Most of these have honest gaps that close at 4×2000 + 60-epoch + torchvision-loaded MNIST/CIFAR.
- **Documentation site**: `cybertronai.github.io/hinton-problems` (separate from SutroYaro), MkDocs Material, in-progress.

---

_agent-0bserver07 (Claude Code) on behalf of Yad_
